connectors/sources/sharepoint.py

```diff
@@ -0,0 +1,715 @@
+        request_headers = {
+            "accept": "application/json",
+            "content-type": "application/json",
+        }
+        timeout = aiohttp.ClientTimeout(total=None)
+
+        if self.is_cloud:
+            basic_auth = None
+            await self._set_access_token()
+        else:
+            basic_auth = aiohttp.BasicAuth(
+                login=self.configuration["username"],
+                password=self.configuration["password"],
+            )
+        return aiohttp.ClientSession(
+            auth=basic_auth,
+            headers=request_headers,
+            timeout=timeout,
+            raise_for_status=True,
+        )
+
+    async def close(self):
+        """Closes unclosed client session"""
+        if self.session is None:
+            return
+        await self.session.close()
+
+    async def _api_call(self, url_name, url="", **url_kwargs):
+        """Make an API call to the Sharepoint Server/Online
+
+        Args:
+            url_name (str): Sharepoint url name to be executed.
+            url(str, optional): Paginated url for drive and list items. Defaults to "".
+            url_kwargs (dict): Url kwargs to format the query.
+        Raises:
+            exception: An instance of an exception class.
+
+        Yields:
+            data: API response.
+        """
+        retry = 0
+        # If pagination happens for list and drive items then next pagination url comes in response which will be passed in url field.
+        if url == "":
+            url = URLS[url_name].format(**url_kwargs)
+        while True:
+            try:
+                async with self.session.get(
+                    url=url,
+                    ssl=self.ssl_ctx,
+                    headers={"Authorization": f"Bearer {self.access_token}"}
+                    if self.is_cloud
+                    else None,
+                ) as result:
+                    if url_name == ATTACHMENT:
+                        yield result
+                    else:
+                        yield await result.json()
+                    break
+            except Exception as exception:
+                if isinstance(
+                    exception, aiohttp.client_exceptions.ClientResponseError
+                ) and "token has expired" in exception.headers.get(
+                    "x-ms-diagnostics", ""
+                ):
+                    await self._set_access_token()
+                elif isinstance(
+                    exception, aiohttp.client_exceptions.ServerDisconnectedError
+                ):
+                    await self.close()
+                    self.session = await self._generate_session()
+                retry += 1
+                if retry > self.retry_count:
+                    raise exception
+                logger.warning(
+                    f"Retry count: {retry} out of {self.retry_count}. Exception: {exception}"
+                )
+                await asyncio.sleep(RETRY_INTERVAL**retry)
```

Comment:
Instead of doing asyncio.sleep, please use CancellableSleeps, see pattern here: https://github.com/elastic/connectors-python/blob/main/connectors/sources/mysql.py#L91